### PR TITLE
Allow security disabling in Upload class for unit testing purposes (#4751)

### PR DESCRIPTION
**Pull request by shadlaws**

### CODE REVIEW

**Comment by shadlaws** [comment]
Issue: http://dev.kohanaframework.org/issues/4751

**Comment by rjd22** [comment]
@shadlaws It would be nice if you could add unittests to kohana/kohana that test the upload class and this functionality. Just to make sure everything works like it's supposed to.

File: classes/Kohana/Upload.php [hunk 1]
```diff
@@ -31,6 +32,15 @@ class Kohana_Upload {
 	public static $default_directory = 'upload';
 
 	/**
+	 * By default, the methods in this class are secure against fake uploads.  However,
+	 * for unit testing purposes, it may be useful to disable security to test functionality
+	 * downstream.  This should never be used in production installations!
+	 *
+	 * @var  boolean  allow security disabling using [Upload::disable_security]
+	 */
+	public static $allow_security_disabling = FALSE;
```
Comment (zeelot) [inline classes/Kohana/Upload.php:41]: I'm not sure we need this kind of flag. Having a property to makes one of the methods do nothing seems a bit odd.

File: classes/Kohana/Upload.php [hunk 2]
```diff
@@ -131,7 +144,7 @@ public static function not_empty(array $file)
 		return (isset($file['error'])
 			AND isset($file['tmp_name'])
 			AND $file['error'] === UPLOAD_ERR_OK
-			AND is_uploaded_file($file['tmp_name']));
+			AND (static::$_disable_security OR is_uploaded_file($file['tmp_name'])));
```
Comment (zeelot) [inline classes/Kohana/Upload.php:147]: Can you use `Upload` instead of `static` to match our conventions?

File: classes/Kohana/Upload.php (hunk 1 above)
Comment (zombor) [inline classes/Kohana/Upload.php:41]: I agree. Adding methods "for testing purposes" is wrong.

**Comment by zeelot** [comment]
And just realized this can't be done in `3.3.x` as adding method and properties is changing the API. I'm going to close this pull request and move the ticket to `3.4.0` but I'll keep a reference to here in redmine so we can find the solution.

**Comment by arteymix** [comment]
To solve that issue, I override Unittest_TestCase and override the Upload class in it.

``` php
class Unittest_TestCase {

}

class Upload extends Kohana_Upload {

    public static function not_empty(array $file)                                                                                                 
    {                                                                                                                                          
        return (isset($file['error']) and isset($file['tmp_name']) and $file['error'] === UPLOAD_ERR_OK);                                         
        // AND is_uploaded_file($file['tmp_name']));                                                                                              
    } 
}
```

I think we could solve this simply by documenting the solution. Adding a flag to disable security enforcment is a really tricky thing.

**Pull request closed without merging**